File: 0727_array/13. Roman to Integer.java
class Solution {
    public int romanToInt(String s) {

        HashMap<Character, Integer> roman = new HashMap<Character, Integer>();
        roman.put('M', 1000);
        roman.put('D', 500);
        roman.put('C', 100);
        roman.put('L', 50);
        roman.put('X', 10);
        roman.put('V', 5);
        roman.put('I', 1);

        int result = 0;

        for (int i = 0; i < s.length() - 1; i++) {
            // 예외의 경우
            if (roman.get(s.charAt(i)) < roman.get(s.charAt(i+1))) {
                result -= roman.get(s.charAt(i));
            }
            else {
                result += roman.get(s.charAt(i));
            }
        }

        // 마지막 수는 항상 더한다.
        return result + roman.get(s.charAt(s.length()-1));

    }
}

File: 0727_array/13. Roman to Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:

        main = {"I": 1,
                "V": 5,
                "X": 10,
                "L": 50,
                "C": 100,
                "D": 500,
                "M": 1000}

        sub = {"IV": 4,
               "IX": 9,
               "XL": 40,
               "XC": 90,
               "CD": 400,
               "CM": 900}

        result, i = 0, 0
        while True:
            if i == len(s):
                break

            if len(s) == 1 or i == len(s)-1:
                result += main[s[i]]
                break

            if s[i] + s[i+1] in sub:
                result += sub[s[i]+s[i+1]]
                i += 2
                continue

            if s[i] in main:
                result += main[s[i]]
                i += 1

        return result

